idseq_dag/steps/run_star.py

```diff
@@ -1,12 +1,218 @@
+import multiprocessing
+import os
+
 from idseq_dag.engine.pipeline_step import PipelineStep
 import idseq_dag.util.command as command
+import idseq_dag.util.s3 as s3
+import idseq_dag.util.log as log
 
 
 class PipelineStepRunStar(PipelineStep):
+    max_input_reads = 75 * 1000 * 1000
+
     def run(self):
-        input_file_0 = self.input_files_local[0][0]
-        input_file_1 = self.input_files_local[0][1]
-        output_file_0 = self.output_files_local()[0]
-        output_file_1 = self.output_files_local()[1]
-        command.execute("gzip -dc %s |head -20 > %s" % (input_file_0, output_file_0))
-        command.execute("gzip -dc %s |head -20 > %s" % (input_file_1, output_file_1))
+        """Run STAR to filter out host reads."""
+        # Setup
+        input_files = self.input_files_local[0][0:2]
+        star_genome = self.additional_files["star_genome"]
+        ref_dir = self.ref_dir_local
+        scratch_dir = os.path.join(self.output_dir_local, "scratch")
+        total_counts_from_star = {}
+        output_files_local = self.output_files_local()
+        num_inputs = len(self.input_files[0])
+
+        # Check genome file. Ex: s3://...STAR_genome.tar -> /ref/STAR_genome
+        basename = os.path.basename(star_genome).rstrip(".tar")
+        genome_dir = os.path.join(ref_dir, basename)
+        assert os.path.exists(genome_dir)
+        assert genome_dir is not None
+
+        # Check parts file for the number of partitioned indexes
+        parts_file = os.path.join(genome_dir, "parts.txt")
+        assert os.path.isfile(parts_file)
+        with open(parts_file, 'rb') as parts_f:
+            num_parts = int(parts_f.read())
+
+        # Run STAR on each partition and save the unmapped read info
+        unmapped = input_files
+        for part_idx in range(num_parts):
+            tmp = f"{scratch_dir}/star-part-{part_idx}"
+            genome_part = f"{genome_dir}/part-{part_idx}"
+            count_genes = part_idx == 0
+            self.run_star_part(tmp, genome_part, unmapped, count_genes)
+
+            unmapped = self.sync_pairs(self.unmapped_files_in(tmp, num_inputs))
+
+            # Run part 0 in gene-counting mode:
+            # (a) ERCCs are doped into part 0 and we want their counts.
+            # (b) If there is only 1 part (e.g. human), the host gene counts also
+            # make sense.
+            # (c) At part 0, we can also extract out total input reads and if the
+            # total_counts is exactly the same as max_input_reads then we know the
+            # input file is truncated.
+            if part_idx == 0:
+                gene_count_file = os.path.join(tmp, "ReadsPerGene.out.tab")
+                self.extract_total_counts(tmp, num_inputs,
+                                          total_counts_from_star)
+                if os.path.isfile(gene_count_file):
+                    basename = os.path.basename(gene_count_file)
```

Comment:
actually, i think we actually move ReadsPerGene.out.tab to reads_per_gene.tab or something like that so instead of using the basename, you can just add an additional attribute called say "output_gene_file" 
and you move genome_count_file to output_dir_local/self.additional_attributes["output_gene_file"], 
you only do it when that attribute exists.